Core Concepts week 1 day 4 update (#33)
microsoft/win-student-devs: docs/30DaysOfPWA/core-concepts/04.md

> # 1.4 Service Workers

Author: ```suggestion
# 1.4 Make PWA Reliable
```

 * _Web App Manifests_ are like app resumes - providing identity, branding and skills information that is needed for app installation (on device) or publishing (to app stores).

**What we'll cover today:** We'll explore the remaining building blocks (HTTPS, Service Workers) with specific focus on [using Service Workers](https://aka.ms/learn-PWA/30Days-1.4/docs.microsoft.com/en-us/microsoft-edge/progressive-web-apps-chromium/how-to/service-workers).

 
---

## PWAs are like startups!

To set the stage, let's use another analogy. We previously talked about how Web App Manifests are like app resumes. Now think about the PWA like a startup - with each technology being a founding team member with a specialized purpose that helps deliver a progressively enhanced experience..

 * The app **page** is the CEO - it drives the core experience and remains responsive to user needs and interactions. 
 * The **Web App Manifest** is the resume - it describes app identity, brand and capabilities to devices and app stores _for installability_.
 * **HTTPS** is the Chief Security Officer (CSO) - it encrypts end-to-end communications between app and server endpoints _for safety_.

>  * The **Service Worker** is the Chief Operations Officer (COO) - it unblocks the CEO from performing time-consuming or asynchronous tasks, and takes proactive actions to ensure reliable operation even when offline.

Reviewer: Continued metaphor is starting to lose me a bit. Feel like it gets a little less relatable/clear as it expands.

Author: Really? I thought this was cool! Resolving as ruled over by Beth :p. JK I don't think we need to change this as people will learn more outside of the analogy anyways.

>  * The **Service Worker** is the Chief Operations Officer (COO) - it unblocks the CEO from performing time-consuming or asynchronous tasks, and takes proactive actions to ensure reliable operation even when offline.

Reviewer 2: I think Nitya means "synchronous" as opposed to async since SWs are async.

> ![Placeholder Banner Only. Replace when final assets ready.](_media/4-service-workers.png)

Author: ```suggestion
![Image describes relationship of PWA startup.](_media/4-service-workers.png)
```

Let's see how this translates to PWA implementations that support desirable traits like safety, network-independent operation, and re-engageability. And in particular, we'll explore service worker implementation and usage, to make this happen.

---

## Make PWAs Safe

> This is achieved with HTTPS.

Reviewer 2: I think this sentence could be dropped altogether.

```suggestion
```

HTTPS enforces end-to-end encryption of all client-server communications, providing privacy safeguards for information exchanged over the network. Visualize this as a Chief Security Officer locking down all entry and access paths into your startup and protecting information from malicious third-party access or tampering.

**HTTPS support is mandatory for using Service Workers**. Thankfully, as we covered [in our earlier post](02.md), it is easy to implement HTTPS support. Use modern cloud hosting providers (who enable it by default) or take advantage of free certificate options (e.g., [Let's Encrypt](https://aka.ms/learn-PWA/30Days-1.4/letsencrypt.com) to secure your own servers.

## Make PWAs Reliable & Re-Engageable

> This is achieved with Service Workers.

Reviewer 2: Don’t feel like this really adds to the content.

```suggestion
```

> Service Workers are a special type of [Web Worker](https://aka.ms/learn-PWA/30Days-1.4/www.w3schools.com/js/js_api_web_workers.asp) that execute long-running or asynchronous tasks in the background, minimizing impact on page performance ("unblocking" the CEO).

Author: I think definition of service worker is clear in the following paragraph. Or do you suggest changing the order of these two?

> Service Workers are a special type of [Web Worker](https://aka.ms/learn-PWA/30Days-1.4/www.w3schools.com/js/js_api_web_workers.asp) that execute long-running or asynchronous tasks in the background, minimizing impact on page performance ("unblocking" the CEO).

Reviewer 2: Can we reference MDN rather than w3schools please? I’ve been actively trying to starve that site of links because it’s often quite bad at doing what it purports to do.

> Service Workers are a special type of [Web Worker](https://aka.ms/learn-PWA/30Days-1.4/www.w3schools.com/js/js_api_web_workers.asp) that execute long-running or asynchronous tasks in the background, minimizing impact on page performance ("unblocking" the CEO).

Author: Yeah no prob

> Service Workers are a special type of [Web Worker](https://aka.ms/learn-PWA/30Days-1.4/www.w3schools.com/js/js_api_web_workers.asp). Web Workers operate in a separate thread, allowing them to execute long-running or asynchronous tasks in the background, minimizing the impact on page performance ("unblocking" the CEO).

Author: ```suggestion
Service Workers are a special type of [Web Worker](https://aka.ms/learn-PWA/30Days-1.4/developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API). Web Workers operate in a separate thread, allowing them to execute long-running or asynchronous tasks in the background, minimizing the impact on page performance ("unblocking" the CEO).
```

Service Workers make PWA operation _reliable_ by helping deliver usable experiences even under flaky or offline network conditions. They do this by _intercepting_ network fetch requests from the page (CEO) and _strategically_ handling them using cached responses (if offline), or network-fetched resources (if real-time), or some combination of both based on predefined caching strategies for resource types.

> Service Workers make PWAs _re-engageable_ by having the ability to alert users to app changes or context, even if the page itself is inactive. They do this by listening for asynchronous push notifications (from server) and working with platform capabilities to deliver alerts in a device-familiar way. When users engage with the alert, they are drawn back into the app seamlessly - just like with other native app experiences.

Reviewer 2: ```suggestion
Service Workers make PWAs _re-engageable_ by having the ability to alert users to app changes or context, even if the page itself is inactive. They do this by listening for asynchronous push notifications (from a server) and working with platform capabilities to deliver alerts in a device-familiar way. When users engage with the alert, they are drawn back into the app seamlessly - just like with other native app experiences.
```

---

## How do Service Workers work?

From a development perspective, we need to know two concepts:
 * Service Worker **Registration** - where CEO "hires" the COO.
 * Service Worker **Lifecycle** - where COO "handles" operational events.

> Let's look at registration first. The Service Worker functionality must be implemented in its own file. The location of that file (relative to the root of the app) defines the _scope_ of its authority. Service Workers can only intercept or manage requests to pages within their scope. _Ideally, placing the file at the root of your app ensures service workers will manage all pages within it_.

Reviewer 2: ```suggestion
Let's look at registration first. Like all Web Workers, the Service Worker must be authored in its own file. The location of that file (relative to the root of the app) defines the _scope_ of its authority. Service Workers can only intercept or manage requests to pages within their scope. _Placing the file at the root of your app ensures your service worker will manage all pages within it_.
```

![Inspect PWA in DevTools](_media/dtt-inspect.png)

If you inspect the application source (in _Elements_ tab) you will find this snippet of code for service worker registration:

```javascript
if('serviceWorker' in navigator) {
    // Register the service worker
    navigator.serviceWorker.register('/sw.js', {
        scope: '/'
    });
}
```

Because Service Worker is a more recent technology that may not be [supported on all browsers](https://aka.ms/learn-PWA/30Days-1.4/caniuse.com/serviceworkers), we test for its existence before registering it. The scope is set implicitly by the file location - the code shows how you can explicitly set this if needed. 

---

## Service Worker: Lifecycle Events

Service worker registration is like onboarding the COO. Once that is complete, the service worker is ready to listen for lifecycle events (install, activate) to set itself up for success. Think of this as three phases:

1. **Registration**: The browser registers the service worker. It is now ready to receive and process events.

> 2. **Installation**: The browser triggers `install` as the first event to the Service Worker. It can use this for pre-caching resources (e.g., populate cache with long-running resources like logos or offline pages) while it waits for older service worker versions to exit.

Reviewer 2: ```suggestion
2. **Installation**: The browser triggers `install` as the first event to the Service Worker. It can use this for pre-caching resources (e.g., populate cache with long-lived resources like logos or offline pages).
```

Reviewer 2: > while it waits for older service worker versions to exit

Not quite. Until the clients are claimed (during activation) or the page is navigated/refreshed, old service workers are still in play.

```javascript
self.addEventListener( "install", function( event ){
    console.log( "WORKER: install event in progress." );
});
```

> 3. **Activation**: The browser sends the `activate` event to indicate that prior service worker versions have exited. This service worker can now do clean up actions (e.g., remove caches from prior version) and ready itself to handle _functional_ events.

Reviewer 2: ```suggestion
3. **Activation**: The browser sends the `activate` event to indicate that the service worker has been installed. This service worker can now do clean up actions (e.g., remove old caches from prior version) and ready itself to handle _functional_ events. If there is an old service worker in play, you can use `clients.claim()` to immediately replace the old service worker with your new one.
```

```javascript
self.addEventListener( "activate", function( event ){

>     console.log( "WORKER: install event in progress." );

Reviewer 2: ```suggestion
    console.log( "WORKER: activation event in progress." );
    clients.claim();
    console.log( "WORKER: all clients are now controlled by me! Mwahahaha!" );
```

Reviewer 2: Alt: "All your clients are belong to me"

});
```

## Service Worker: Functional Events

Functional events are those that require the asynchronous or background processing abilities of service workers to support reliable and re-enageable behaviors. For now, think about just two: "fetch" and "push".

1. The [**fetch event**](https://aka.ms/learn-PWA/30Days-1.4/developer.mozilla.org/en-US/docs/Web/API/FetchEvent) is triggered when the browser tries to access a page that lies within the _scope_ of the service worker. The service worker acts as an interceptor - returning a response either from the cache or from the network (or some combination of both) based on predefined strategies. We'll cover more on this in the next post.

```javascript
self.addEventListener( "fetch", function( event ){
    console.log( "WORKER: Fetching", event.request );
});
```

> 2. The [**push event**](https://aka.ms/learn-PWA/30Days-1.4/developer.mozilla.org/en-US/docs/Web/API/PushEvent) is triggered when the browser receives a push message from a server to display as a toast notification to users. This occurs only if the PWA had previously _subscribed_ for server notifications _and_ user had granted PWA permissions to receive them. Push events are critical to [re-engaging users](https://aka.ms/learn-PWA/30Days-1.4/docs.microsoft.com/en-us/microsoft-edge/progressive-web-apps-chromium/how-to/notifications-badges) when the app is not otherwise active.

Reviewer 2: ```suggestion
2. The [**push event**](https://aka.ms/learn-PWA/30Days-1.4/developer.mozilla.org/en-US/docs/Web/API/PushEvent) is triggered when the browser receives a push message from a server to display as a toast notification to users. This occurs only if the PWA had previously _subscribed_ for server notifications _and_ user has granted the PWA permission to receive them. Push events are critical to [re-engaging users](https://aka.ms/learn-PWA/30Days-1.4/docs.microsoft.com/en-us/microsoft-edge/progressive-web-apps-chromium/how-to/notifications-badges) when the app is not otherwise active.
```
